File: 1912.design-movie-rental-system.py
# @leet imports start
from typing import *

# @leet imports end

# @leet start
from heapq import heappush, heappop, heapify

# NOTE: lol ignore this one, see the java.
# Lazy delete is annoying


class MovieRentingSystem:

    def __init__(self, n: int, entries: List[List[int]]):
        # entries: list of (shop, movie, price)

        # minheap of (price, movie, shop)
        self.unrented = [(price, movie, shop) for (shop, movie, price) in entries]
        heapify(self.unrented)

        # minheap of (price, shop, movie)
        self.rented = []

        # (shop, movie) = price. where +shop if not rented, -shop if rented
        self.entries = dict()

        for shop, movie, price in entries:
            self.entries[(shop, movie)] = price

    def search(self, movie: int) -> List[int]:
        res = []
        temp = []
        while self.unrented and len(res) < 5:
            price, m, shop = heappop(self.unrented)
            if m != movie or (shop, m) not in self.entries:
                temp.append((price, m, shop))
                continue
            res.append((price, m, shop))
        for x in res + temp:
            heappush(self.unrented, x)
        return [shop for (_, _, shop) in res]

    def rent(self, shop: int, movie: int) -> None:
        price = self.entries[(shop, movie)]
        del self.entries[(shop, movie)]
        self.entries[(-shop, movie)] = price
        heappush(self.rented, (price, shop, movie))

    def drop(self, shop: int, movie: int) -> None:
        price = self.entries[(-shop, movie)]
        del self.entries[(-shop, movie)]
        self.entries[(shop, movie)] = price
        heappush(self.unrented, (price, movie, shop))

    def report(self) -> List[List[int]]:
        res = []
        temp = []
        while self.rented and len(res) < 5:
            price, shop, movie = heappop(self.rented)
            if (-shop, movie) not in self.entries:
                temp.append((price, shop, movie))
                continue
            res.append((price, shop, movie))
        for x in res + temp:
            heappush(self.rented, x)
        return [[shop, movie] for (_, shop, movie) in res]


# Your MovieRentingSystem object will be instantiated and called as such:
# obj = MovieRentingSystem(n, entries)
# param_1 = obj.search(movie)
# obj.rent(shop,movie)
# obj.drop(shop,movie)
# param_4 = obj.report()
# @leet end

File: 1912.design-movie-rental-system.java
// @leet imports start

import java.util.*;
import java.math.*;
// @leet imports end

// @leet start
class MovieRentingSystem {
    private record Entry(int price, int shop, int movie) {
    }

    private record Pair(int shop, int movie) {
    }

    Comparator<Entry> comparator = (o1, o2) -> {
        if (o1.price != o2.price)
            return Integer.compare(o1.price, o2.price);
        if (o1.shop != o2.shop)
            return Integer.compare(o1.shop, o2.shop);
        return Integer.compare(o1.movie, o2.movie);
    };

    // movieId -> TreeSet of Entries
    Map<Integer, Set<Entry>> unrented = new HashMap<>();
    // (shop, movie) -> price
    Map<Pair, Integer> price = new HashMap<>();
    Set<Entry> rented = new TreeSet<>(comparator);

    public MovieRentingSystem(int n, int[][] entries) {
        for (int[] entry : entries) {
            int s = entry[0], m = entry[1], p = entry[2];
            unrented.computeIfAbsent(m, x -> new TreeSet<>(comparator))
                    .add(new Entry(p, s, m));
            price.put(new Pair(s, m), p);
        }
    }

    public List<Integer> search(int movie) {
        return unrented.getOrDefault(movie, Collections.emptySet())
                .stream()
                .limit(5)
                .map(e -> e.shop)
                .toList();
    }

    public void rent(int shop, int movie) {
        int p = price.get(new Pair(shop, movie));
        unrented.get(movie).remove(new Entry(p, shop, movie));
        rented.add(new Entry(p, shop, movie));
    }

    public void drop(int shop, int movie) {
        int p = price.get(new Pair(shop, movie));
        unrented.get(movie).add(new Entry(p, shop, movie));
        rented.remove(new Entry(p, shop, movie));
    }

    public List<List<Integer>> report() {
        return rented.stream()
                .limit(5)
                .map(e -> List.of(e.shop, e.movie))
                .toList();
    }

}

/**
 * Your MovieRentingSystem object will be instantiated and called as such:
 * MovieRentingSystem obj = new MovieRentingSystem(n, entries);
 * List<Integer> param_1 = obj.search(movie);
 * obj.rent(shop,movie);
 * obj.drop(shop,movie);
 * List<List<Integer>> param_4 = obj.report();
 */
// @leet end
